packages/core/admin/utils/plugins.js

```diff
@@ -0,0 +1,217 @@
+   * or it's an actual npm module that's installed and resolved via node_modules.
+   *
+   * We first check if the plugin is local to the strapi app, using a regular `resolve` because
+   * the pathToPlugin will be relative i.e. `/Users/my-name/strapi-app/src/plugins/my-plugin`.
+   *
+   * If the file doesn't exist well then it's probably a node_module, so instead we use `require.resolve`
+   * which will resolve the path to the module in node_modules. If it fails with the specific code `MODULE_NOT_FOUND`
+   * then it doesn't have an admin part to the package.
+   *
+   * NOTE: we should try to move to `./package.json[exports]` map with bundling of our own plugins,
+   * because these entry files are written in commonjs restricting features e.g. tree-shaking.
+   */
+  try {
+    const isLocalPluginWithLegacyAdminFile = fs.existsSync(
+      path.resolve(`${plugin.pathToPlugin}/strapi-admin.js`)
+    );
+
+    if (!isLocalPluginWithLegacyAdminFile) {
+      const isModuleWithFE = require.resolve(`${plugin.pathToPlugin}/strapi-admin`);
+
+      return isModuleWithFE;
+    }
+
+    return isLocalPluginWithLegacyAdminFile;
+  } catch (err) {
+    if (err.code === 'MODULE_NOT_FOUND') {
+      /**
+       * the plugin does not contain FE code, so we
+       * don't want to import it anyway
+       */
+      return false;
+    }
+
+    throw err;
+  }
+};
+
+/**
+ *
+ * @param {Plugin[]} plugins
+ * @param {string} dest
+ * @returns {void}
+ */
+async function createPluginFile(plugins, dest) {
+  const pluginsArray = plugins.map(({ pathToPlugin, name, info }) => {
+    const shortName = camelCase(name);
+
+    let realPath = '';
+
+    /**
+     * We're using a module here so we want to keep using the module resolution procedure.
+     */
+    if (info?.packageName || info?.required) {
+      /**
+       * path.join, on windows, it uses backslashes to resolve path.
+       * The problem is that Webpack does not windows paths
+       * With this tool, we need to rely on "/" and not "\".
+       * This is the reason why '..\\..\\..\\node_modules\\@strapi\\plugin-content-type-builder/strapi-admin.js' was not working.
+       * The regexp at line 105 aims to replace the windows backslashes by standard slash so that webpack can deal with them.
+       * Backslash looks to work only for absolute paths with webpack => https://webpack.js.org/concepts/module-resolution/#absolute-paths
+       */
+      realPath = path.join(pathToPlugin, 'strapi-admin').replace(/\\/g, '/');
+    } else {
+      realPath = path
+        .join(path.relative(path.resolve(dest, 'admin', 'src'), pathToPlugin), 'strapi-admin')
+        .replace(/\\/g, '/');
+    }
+
+    return {
+      name,
+      pathToPlugin: realPath,
+      shortName,
+    };
+  });
+
+  const content = `
+${pluginsArray
+  .map(({ pathToPlugin, shortName }) => {
+    const req = `'${pathToPlugin}'`;
+
+    return `import ${shortName} from ${req};`;
```

Comment:
It would be nice to add some validation on shortName and pathToPlugin.

It shouldn't be a problem in theory, but if someone has strapi.name missing or with invalid characters (which we don't allow on generation, but someone could have edited it) it would be better to catch it and throw an error rather than generating code that is broken (and even if extreeeeeemely unlikely, potentially dangerous) rather than leading them to debug.

Reply:
Yeah, just something simple to make sure it exists and only contains expected characters. But like you said, it should already be adhering to rules, so this is just a minor point specifically to catch edges we haven't thought of.